### PR TITLE
reverse string

### LINKED ISSUE #20: reverse string
Given a string, return the string reversed.

### PR DESCRIPTION
**Pull request by Nada-saad635**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/reverse_string.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+def reverse_string(input_string: str) -> str:
```
Comment (SiSaR-Pal) [inline solutions/reverse_string.py:26]: Add a module-level docstring  to satisfy the requirement for a module docstring.

**Review by SiSaR-Pal — COMMENTED** [review]
The provided code and tests are well-written and comprehensive, adhering to Python best practices and meeting the checklist criteria. No significant changes are necessary.

File: solutions/tests/test_reverse_string.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        self.assertEqual(reverse_string("madam"), "madam")
+
+    def test_numbers_in_string(self):
+        """Test reversing a string containing numbers."""
+        self.assertEqual(reverse_string("12345"), "54321")
+
+    def test_special_characters(self):
+        """Test reversing a string with special characters."""
+        self.assertEqual(reverse_string("!@#$"), "$#@!")
+
```
Comment (heba3) [inline solutions/tests/test_reverse_string.py:40]: The function looks great and is easy to follow! One small suggestion: for very long strings, this approach might use a lot of memory because it creates a new reversed sequence.

Comment (heba3) [reply solutions/tests/test_reverse_string.py:40]:
```python
def test_very_long_string(self):
    """Test reversing a very long string."""
    long_string = "a" * 1000  # A string with 1000 'a's
    self.assertEqual(reverse_string(long_string), "a" * 1000)

File: solutions/reverse_string.py (hunk 1 above)
Comment (Nada-saad635) [reply solutions/reverse_string.py:26]: done can you check it out ?

File: solutions/tests/test_reverse_string.py (hunk 2 above)
Comment (Nada-saad635) [reply solutions/tests/test_reverse_string.py:40]: what you suggest to do ?

File: solutions/tests/test_reverse_string.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
+import unittest
+from ..reverse_string import reverse_string
+
+"""
```
Comment (Elshikh-Amro) [inline solutions/tests/test_reverse_string.py:1]: i think the docstring should be above the imports or any code

**Review by Elshikh-Amro — COMMENTED** [review]
goood jooob!!!!!
this is goood and very well written

**Review by SiSaR-Pal — DISMISSED** [review]

File: solutions/tests/test_reverse_string.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+    def test_numbers_in_string(self):
+        """Test reversing a string containing numbers."""
+        self.assertEqual(reverse_string("12345"), "54321")
+
+    def test_special_characters(self):
+        """Test reversing a string with special characters."""
+        self.assertEqual(reverse_string("!@#$"), "$#@!")
+
+    def test_non_string_input(self):
```
Comment (omniaNS) [inline solutions/tests/test_reverse_string.py:41]: according to the checklist, each unit test should have only one assertion

File: solutions/tests/test_reverse_string.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        """Test that non-string input raises an AssertionError."""
+        with self.assertRaises(AssertionError):
```
Comment (omniaNS) [inline solutions/tests/test_reverse_string.py:48]: according to the checklist, every unit test should have only one assertion, i suggest separating your 3 assertions into 3 unit tests

**Review by omniaNS — CHANGES_REQUESTED** [review]

File: solutions/tests/test_reverse_string.py (hunk 5 above)
Comment (Nada-saad635) [reply solutions/tests/test_reverse_string.py:48]: done with this

File: solutions/tests/test_reverse_string.py (hunk 3 above)
Comment (Nada-saad635) [reply solutions/tests/test_reverse_string.py:1]: done with it

**Review by omniaNS — DISMISSED** [review]

File: solutions/tests/test_reverse_string.py (hunk 2 above)
Comment (Nada-saad635) [reply solutions/tests/test_reverse_string.py:40]: done i add it !

**Review by heba3 — APPROVED** [review]

**Review by SiSaR-Pal — APPROVED** [review]
